File: packages/ld-adapter/src/write/apply-targeting.ts
import { ApplyTargetingInputSchema, type TargetingIntent } from '@ff-promo/contracts';
import type { FlagState } from '@ff-promo/contracts';
import {
	SEMANTIC_PATCH_CONTENT_TYPE,
	type LaunchDarklyRawClient,
} from '../client/ld-api-client.js';
import type { RateLimitedLdClient } from '../client/rate-limited-client.js';
import { getFlagState } from '../read/get-flag-state.js';
import { resolveRuleId } from '../resolve/rule-resolver.js';
import { resolveVariationId } from '../resolve/variation-resolver.js';
import {
	buildTargetingPatchBody,
	type ResolvedRolloutIds,
} from './semantic-patch.js';

export type ApplyTargetingDeps = {
	rateLimitedClient: RateLimitedLdClient;
};

class HttpResponseError extends Error {
	readonly status: number;
	readonly response: {
		header: Record<string, string | string[] | undefined>;
		body?: unknown;
	};

	constructor(
		status: number,
		message: string,
		body?: unknown,
		headers?: Record<string, string | string[] | undefined>,
	) {
		super(message);
		this.name = 'HttpResponseError';
		this.status = status;
		this.response = { body, header: headers ?? {} };
	}
}

async function semanticPatchFeatureFlag(
	client: LaunchDarklyRawClient,
	projectKey: string,
	flagKey: string,
	body: Record<string, unknown>,
): Promise<unknown> {
	const url = `${client.config.baseUrl}/api/v2/flags/${encodeURIComponent(projectKey)}/${encodeURIComponent(flagKey)}`;
	const response = await fetch(url, {
		method: 'PATCH',
		headers: {
			Authorization: client.config.accessToken,
			'LD-API-Version': client.config.apiVersion,
			'Content-Type': SEMANTIC_PATCH_CONTENT_TYPE,
			Accept: 'application/json',
		},
		body: JSON.stringify(body),
	});

	const responseHeaders: Record<string, string> = {};
	for (const [key, value] of response.headers.entries()) {
		responseHeaders[key] = value;
	}

	const responseBody = await response.json().catch(() => undefined);
	if (!response.ok) {
		throw new HttpResponseError(
			response.status,
			`LaunchDarkly PATCH failed with status ${response.status}`,
			responseBody,
			responseHeaders,
		);
	}

	return responseBody;
}

export async function applyTargeting(
	deps: ApplyTargetingDeps,
	input: {
		projectKey: string;
		flagKey: string;
		intent: TargetingIntent;
	},
): Promise<FlagState> {
	const parsed = ApplyTargetingInputSchema.parse(input);
	const { rateLimitedClient } = deps;
	const { rawClient } = rateLimitedClient;

	const flagState = await rateLimitedClient.schedule(() =>
		getFlagState({ client: rawClient }, {
			projectKey: parsed.projectKey,
			flagKey: parsed.flagKey,
			environmentKey: parsed.intent.environmentKey,
		}),
	);

	let resolved: ResolvedRolloutIds | undefined;
	if (parsed.intent.rollout) {
		const rollout = parsed.intent.rollout;
		resolved = {
			treatmentVariationId: resolveVariationId(
				flagState,
				rollout.treatmentVariationRef,
			),
			controlVariationId: resolveVariationId(
				flagState,
				rollout.controlVariationRef,
			),
		};
		if (rollout.mode === 'rule') {
			if (!rollout.ruleRef) {
				throw new Error('ruleRef required when rollout.mode is rule');
			}
			resolved.ruleId = resolveRuleId(
				flagState,
				parsed.intent.environmentKey,
				rollout.ruleRef,
			);
		}
	}

	const patchBody = buildTargetingPatchBody(parsed.intent, resolved);

	await rateLimitedClient.schedule(() =>
		semanticPatchFeatureFlag(
			rawClient,
			parsed.projectKey,
			parsed.flagKey,
			patchBody,
		),
	);

	return rateLimitedClient.schedule(() =>
		getFlagState({ client: rawClient }, {
			projectKey: parsed.projectKey,
			flagKey: parsed.flagKey,
			environmentKey: parsed.intent.environmentKey,
		}),
	);
}

File: kotlin/modules/ld-adapter/src/main/kotlin/com/ffpromo/ldadapter/write/ApplyTargeting.kt
package com.ffpromo.ldadapter.write

import com.ffpromo.contracts.ApplyTargetingInput
import com.ffpromo.contracts.FlagState
import com.ffpromo.contracts.RolloutIntent
import com.ffpromo.ldadapter.client.LdApiClient
import com.ffpromo.ldadapter.client.RateLimitedLdClient
import com.ffpromo.ldadapter.read.getFlagState
import com.ffpromo.ldadapter.read.GetFlagStateDeps
import com.ffpromo.ldadapter.resolve.resolveRuleId
import com.ffpromo.ldadapter.resolve.resolveVariationId

data class ApplyTargetingDeps(
    val rateLimitedClient: RateLimitedLdClient,
)

suspend fun applyTargeting(deps: ApplyTargetingDeps, input: ApplyTargetingInput): FlagState {
    val rateLimitedClient = deps.rateLimitedClient
    val rawClient = rateLimitedClient.rawClient
    val apiClient = LdApiClient(rawClient)

    val flagState = rateLimitedClient.schedule {
        getFlagState(
            GetFlagStateDeps(rawClient),
            com.ffpromo.contracts.GetFlagStateInput(
                projectKey = input.projectKey,
                flagKey = input.flagKey,
                environmentKey = input.intent.environmentKey,
            ),
        )
    }

    val resolved = resolveRolloutIds(flagState, input.intent.environmentKey, input.intent.rollout)
    val patchBody = buildTargetingPatchBody(input.intent, resolved)

    rateLimitedClient.schedule {
        apiClient.patchFeatureFlag(input.projectKey, input.flagKey, patchBody)
    }

    return rateLimitedClient.schedule {
        getFlagState(
            GetFlagStateDeps(rawClient),
            com.ffpromo.contracts.GetFlagStateInput(
                projectKey = input.projectKey,
                flagKey = input.flagKey,
                environmentKey = input.intent.environmentKey,
            ),
        )
    }
}

private fun resolveRolloutIds(
    flagState: FlagState,
    environmentKey: String,
    rollout: RolloutIntent?,
): ResolvedRolloutIds? {
    if (rollout == null) return null

    val resolved = ResolvedRolloutIds(
        treatmentVariationId = resolveVariationId(flagState, rollout.treatmentVariationRef),
        controlVariationId = resolveVariationId(flagState, rollout.controlVariationRef),
    )

    return if (rollout.mode == "rule") {
        val ruleRef = rollout.ruleRef ?: error("ruleRef required when rollout.mode is rule")
        resolved.copy(ruleId = resolveRuleId(flagState, environmentKey, ruleRef))
    } else {
        resolved
    }
}
